File: java/src/main/java/com/urlshortener/dal/repositories/shorturl/ShortUrlRepository.java
package com.urlshortener.dal.repositories.shorturl;

import com.urlshortener.dal.entities.ShortUrl;
import lombok.NonNull;

import javax.validation.Valid;
import java.util.List;

public interface ShortUrlRepository {
    ShortUrl merge(@NonNull @Valid ShortUrl shortUrl);

    void deleteById(long id, long userId);

    Long countByShortName(@NonNull String shortName);

    ShortUrl findByShortName(@NonNull String shortName);

    ShortUrl findByIdAndUserId(long id, long userId);

    List<ShortUrl> findAllForUser(long userId);
}

File: kotlin/src/main/kotlin/com/urlshortener/dal/repositories/shorturl/ShortUrlRepository.kt
package com.urlshortener.dal.repositories.shorturl

import com.urlshortener.dal.entities.ShortUrl
import javax.validation.Valid

interface ShortUrlRepository {
    fun merge(@Valid shortUrl: ShortUrl): ShortUrl
    fun deleteById(id: Long, userId: Long)
    fun countByShortName(shortName: String): Long
    fun findByShortName(shortName: String): ShortUrl?
    fun findByIdAndUserId(id: Long, userId: Long): ShortUrl?
    fun findAllForUser(userId: Long): List<ShortUrl>
}